### PR TITLE
fix: resolve issues related to upgrading nx >= 16

### LINKED ISSUE #379: [BUG] Error resolving versions of dependencies in a monorepo
### Describe the bug

What initially started out as projen issue seems to affect `nx-monorepo` as well: https://github.com/projen/projen/issues/2641

When running `npx projen` I'm getting the following error as a result of a recent `projen` upgrade:

```
When running npx projen in my mono repo I'm getting the following error:

foo main• ↑4 ❱ npx projen                                         devenv-shell-env
👾 default | ts-node --project tsconfig.dev.json .projenrc.ts
👾 unable to resolve version for api-typescript-runtime from installed modules
👾 unable to resolve version for api-typescript-infra from installed modules
👾 unable to resolve version for api-typescript-runtime from installed modules
```

@mrgrain said the following:

> It pretty much comes down to this line:
> https://github.com/aws/aws-prototyping-sdk/blob/cf57586f2567a571c734606d6f003f6a5051bfca/packages/nx-monorepo/src/nx-monorepo.ts#L766
> 
> projen core now (correctly) expects that subproject dependencies are installed before resolveDepsAndWritePackageJson is called. Overwriting installDependencies on the subproject has been the go to way for monorepos to do this. But of course there's a reason that method is private.
> 
> I think what this highlights is the missing support for workspace-based monorepos in NodeProject.
> I wonder what the API for this should look like.

### Expected Behavior

The dependencies specified in subprojects being resolved.

### Current Behavior

The packages aren't available because the versions can't be resolved.

### Reproduction Steps

Create a new nx-monorepo

```
mkdir repo && npx projen new --from @aws-prototyping-sdk/nx-monorepo
```

Change the `project` variable in `.projenrc.ts` to this:

```typescript
const project = new NxMonorepoProject({
  defaultReleaseBranch: "main",
  devDeps: ["@aws-prototyping-sdk/nx-monorepo", "@aws-prototyping-sdk/type-safe-api"],
  name: "foo",
  projenrcTs: true,

  deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,  /* The "name" in package.json. */
});
```

Then run `npx projen`.

Then change the .projenrc.ts to:

```typescript
import { NxMonorepoProject } from "@aws-prototyping-sdk/nx-monorepo";
import { AwsCdkTypeScriptApp } from "projen/lib/awscdk";
import {
  DocumentationFormat,
  Language,
  ModelLanguage,
  TypeSafeApiProject,
} from "@aws-prototyping-sdk/type-safe-api";

const project = new NxMonorepoProject({
  defaultReleaseBranch: "main",
  devDeps: ["@aws-prototyping-sdk/nx-monorepo", "@aws-prototyping-sdk/type-safe-api"],
  name: "foo",
  projenrcTs: true,

  deps: [],                /* Runtime dependencies of this module. */
  // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
  // packageName: undefined,  /* The "name" in package.json. */
});


/** Contains all the smithy model files and codegenerates documentation, clients and API types.
 */
const api = new TypeSafeApiProject({
  name: "api",
  parent: project,
  outdir: "packages/api",
  runtime: {
    languages: [Language.TYPESCRIPT],
  },
  infrastructure: {
    language: Language.TYPESCRIPT,
  },
  documentation: {
    formats: [DocumentationFormat.HTML_REDOC],
  },
  model: {
    language: ModelLanguage.SMITHY,
    options: {
      smithy: {
        serviceName: {
          namespace: "foo.api",
          serviceName: "FooApi",
        },
      },
    },
  },
});

/** Our CDK based project that will deploy our API on AWS Lambda.
 * Also contains all the code for the Lambda handlers.
 */
new AwsCdkTypeScriptApp({
  name: "infra",
  parent: project,
  outdir: "packages/infra",
  defaultReleaseBranch: "main",
  cdkVersion: "2.76.0",
  deps: [
    api.runtime.typescript!.package.packageName,
    api.infrastructure.typescript!.package.packageName,
    "node-fetch",
  ],
  devDeps: [
  ],
  eslint: false,
  gitignore: ["edgeql-js"],
});

project.synth();
```

And run `npx projen` again. It should fail to resolve the dependencies of the `AwsCdkTypeScriptApp`

### Possible Solution

I'm not sure, but this seems like it might require an API change for `NodeProject`s.

### Additional Information/Context

See https://github.com/projen/projen/issues/2641 and the comments by mrgrain that highlight the way `projen` dealt with version resolution before and after the latest version release.

### PDK version used

0.17.4

### What languages are you seeing this issue on?

Typescript

### Environment details (OS name and version, etc.)

MacOS ARM64 13.3.1

### PR DESCRIPTION
**Pull request by agdimech**
Resolve issues related to upgrading nx >= 16

- Task execution order is now correct due to targetDefaults being set correctly
- Updates to deprecated @nrwl packages
- Pin nx as a peer dep to >= 16
- Package versions are now resolved correctly due to recent change in Projen.
- Removed uber package and refactored samples to take dependencies on individual packages

BREAKING CHANGE: Deprecate uber aws-prototyping-sdk package

Fixes #379, #381, #347

### CODE REVIEW

File: nx.json [hunk 1]
```diff
@@ -40,34 +40,16 @@
...
-        "projects": "dependencies"
-      }
-    ],
     "install": [
       {
         "target": "install",
         "projects": "dependencies"
       }
     ]
   },
```
Comment (JeremyJonas) [inline nx.json]: Should this be moved to `targetDefaults` as well?

File: nx.json [hunk 2]
```diff
@@ -40,34 +40,16 @@
       ]
     }
   },
+  "affected": {
+    "defaultBase": "mainline"
+  },
   "targetDependencies": {
```
Comment (JeremyJonas) [inline nx.json]: This should be removed right? It has been deprecated?

File: packages/nx-monorepo/src/nx-monorepo.ts [hunk 3]
```diff
@@ -451,15 +450,11 @@ export class NxMonorepoProject extends TypeScriptProject {
...
-              target: "build",
-              projects: "dependencies",
-            },
-          ],
-          ...(this.nxConfig?.targetDependencies || {}),
+        targetDefaults: {
+          build: {
+            dependsOn: ["^build"],
+          },
+          ...(this.nxConfig?.targetDefaults || {}),
```
Comment (JeremyJonas) [inline packages/nx-monorepo/src/nx-monorepo.ts]: FYI: `|| {}` is redundant for object spread, only necessary for array spread.

File: packages/nx-monorepo/src/nx-monorepo.ts [hunk 4]
```diff
@@ -776,8 +773,13 @@ export class NxMonorepoProject extends TypeScriptProject {
...
     ) {
-      // @ts-ignore - `installDependencies` is private
-      this.package.installDependencies();
+      try {
+        // @ts-ignore - `installDependencies` is private
+        this.package.installDependencies();
+      } finally {
+        // @ts-ignore - `resolveDepsAndWritePackageJson` is private
+        this.package.resolveDepsAndWritePackageJson();
+      }
```
Comment (JeremyJonas) [inline packages/nx-monorepo/src/nx-monorepo.ts:782]: Missing a `catch` here?

This logic will always run `installDependencies` and `resolveDepsAndWritePackageJson`, even if `installDependencies` fails, but the closure will still fail afterwards if `installDependencies` fails.

I think you want an empty catch here.

**Review by JeremyJonas — APPROVED** [review]
Awesome, thank you. Approved with comments, which if are as expected can merge.

File: nx.json (hunk 2 above)
Comment (agdimech) [reply nx.json]: I don't believe it has as per: https://nx.dev/reference/nx-json

File: nx.json (hunk 1 above)
Comment (agdimech) [reply nx.json]: Ahh yep - good catch

File: packages/nx-monorepo/src/nx-monorepo.ts (hunk 3 above)
Comment (agdimech) [reply packages/nx-monorepo/src/nx-monorepo.ts]: Nice.

File: packages/nx-monorepo/src/nx-monorepo.ts (hunk 4 above)
Comment (agdimech) [reply packages/nx-monorepo/src/nx-monorepo.ts:782]: It's actually correct as it - I want to throw but need to make sure the package.json is written out so it doesn't end up in a corrupt state.

**Pull request merged**